lib/spack/spack/spec.py

```diff
@@ -698,6 +698,19 @@ def _headers_default_handler(descriptor, spec, cls):
         raise RuntimeError(msg.format(spec.name, spec.prefix.include))
 
 
+def _all_headers_default_handler(descriptor, spec, cls):
```

Comment:
The idea behind `ForwardQueryToPackage` was to have a dispatch mechanism designed to:

1. let each package reply for the part of information that is under its control
2. permit virtual packages to reply in different ways, according to how they were queried (see the tutorial [here](https://spack.readthedocs.io/en/latest/tutorial_advanced_packaging.html#retrieving-library-information) or [here](https://spack.readthedocs.io/en/latest/tutorial_advanced_packaging.html#attach-attributes-to-other-packages) for more info)
3. permit to customize queries, if a package provides more than one API or library (see `hdf5.libs` methods for an example)

I think this defeats at least two purposes, as:

1. the information it returns is non-local (we cycle through all dependencies)
3. it's not clear who should handle extra query parameters (a single package? all packages? what if we have clashes on parameters name?)

This is not to say that the PR is a bad idea, but imo it would be better to have a convenience function to compute something like this rather than relying on the `ForwardToQuery` descriptor.

p.s. another thing that reinforces my thoughts above is that I see no reason for a package to override `all_headers` or `all_libs` and provide a different behavior than the default one. Perhaps you have a counterexample?

Reply:
The reason the implementation of `.all_headers` and `.all_libs` uses the exact same implementation mechanism (in addition to the simplicity) is that all the things one can do with `.headers` and `.libs` are also the things that one wants to be able to do with `.all_headers` and `.all_libs`:
1. for any non-virtual package `'x'`, one can get its libraries `spec['x'].libs` (only the libraries that are part of the package `'x'`); often, in addition to that, one will want to be able to get `spec['x'].all_libs` - all libraries from `spec['x'].libs` and all the libraries they depend on
2. lets assume that package `'x'` depends directly on a package `'y'`; the libraries in `'x'` may depend only on a subset of the libraries contained in `'y'` and that information is only known (and should be provided) by package `'x'`; in other words only `'x'` knows that, for example, `spec['x'].all_libs = spec['x'].libs + spec['y+a+b'].all_libs` where `spec['y+a+b'].all_libs` is a query to package `'y'` for its libraries `'ya'` and `'yb'` with all other libraries they depend on (package `'y'` may also include other libraries `'yc'`, `'yd'`, etc.)
3. if package `'x'` has two libraries, `'xa'` and `'xb'` and we want to query `spec['x+a'].all_libs` the direct dependency of `'xa'` on the libraries in `'y'` is known only by `'x'` itself, e.g. `spec['x+a'].all_libs = spec['x+a'].libs + spec['y+b'].all_libs`
4. any virtual packages provided by `'x'`, say `'A'` and `'B'` represent subsets of the libraries in `'x'`, e.g. `spec['A'].libs = spec['x+a+b'].libs`, assuming that `'A'` is resolved through `'x'`; in this case it is again `'x'` that has to provide this information through its `.A_libs` property; similarly, `spec['A'].all_libs` will be provided by the `.A_all_libs` property of `'x'` (which in the case `spec['A'].libs = spec['x+a+b'].libs` will be given by `spec['A'].all_libs = spec['x+a+b'].all_libs`).

The proposed default handler of the `.all_libs` property implements an overestimate of the exact resolutions (if we make one assumption, discussed below). The packages may choose to make the resolutions precise by implementing `.all_libs`, and for the provided virtual packages - by implementing `.A_all_libs`, `.B_all_libs`, etc. To be precise, the implementations of these properties need to handle queries like `spec['x+a'].all_libs`, `spec['x+a+b'].all_libs`, etc. The proposed default implementation always returns an overestimate, i.e. a set containing all necessary dependencies plus potentially some redundant dependencies. This may not be optimal but should always work. The one assumption we need to make (in order to guarantee the overestimate with the default handler) is that the query `spec['x'].libs` returns all libraries in package `'x'`. I think `'hdf5'`, and probably other packages, do not conform to this rule, so a special query may be needed, e.g. `spec['x+all'].libs` that should return all libraries build for the given spec of package `'x'`.

This default implementation can be improved to make it a bit more precise in the case when some of the packages (that the queried package depends on) implement `.all_libs` with a more accurate dependency resolution. This implementation will need to traverse only the direct package dependencies (of the queried package) and then call their `.all_libs` properties. If such a traversal is easy to perform, it will be very easy to make this improvement.

I think that this PR will provide a simple extension to spack that will be useful to many users writing spack packages.

Reply:
What I was thinking, based on your use case, was something like (function not tested within Spack):
```python
def all_libs(spec, specifiers=None):
    """Combines the ``libs`` attribute of spec and its dependencies.

    If the ``specifiers`` dictionary is present, it will be used to refine the queries.

    Args:
        spec (Spec): spec for which we want all the libraries combined together
        specifiers (dict): optional dictionary containing query arguments. The special
            value None can be used to skip     
    """    

    libs = LibraryList()
    for s in spec.traverse(root=True):
        query = ''

        # TODO: need to handle virtual specs ?
        if s.name in specifiers:
           query = specifiers[s.name]
           # Skip this spec if explicitly told to do so
           if query is None:
               continue
        
        tag = ':'.join([s.name, query]) if query else s.name

        libs = libs + s[tag].libs
    
    return libs
```
More in details on your comments:
 
> all the things one can do with .headers and .libs are also the things that one wants to be able to do with .all_headers and .all_libs

This is tied to the (duck) type that is returned by the property, or function, or method. In that sense the function above should satisfy this requirement and point 1. you raise above.

Point 2. can be already solved in a simpler way, where `y` takes care of returning the correct libraries based on its variants (pseudo-code below):

```python
# Package X
class X(Package):
    depends_on('y+a+b')

    ...

#Package Y
class Y(Package):

    @property
    def libs(self):
        libraries = []

        if '+a' in self.spec:
           libraries += ['libya']

        if '+b' in self.spec:
           libraries += ['libyb']


        return ...

# Client that needs the list of all libraries
x = Spec('x')
libs = all_libs(x)
```

I guess point 3. also falls in the above. Also, shame on me, but I don't undestand what you mean by point 4. :smile:

Another point I think is relevant: do you have compelling use cases for point 2. 3. and 4. above? At least for 2. and 3. I can see the theoretical problem, but cannot think of real cases where we need something like that.

Reply:
I agree that point 1 can be handled with the function you propose. However points 2, 3, and 4 cannot, if you want to have precise resolution. The reason is (in your example for point 2) that the dependency of `'x'`, `depends_on('y+a+b')`, can be resolved with `^y+a+b+c` (e.g. requested by a user on the command line) and then `Y.libs` will return `['libya','libyb','libyc']` which is not wrong but `'libyc'` is not really necessary. It is also common for packages to not expose every library they build as a separate variant, e.g. `suite-sparse`, so variants do not necessarily reflect precise library dependencies.

Regarding point 4: I was just trying to say that virtual packages provided by a package are similar to point 3 - they represent a query for a subset of libraries that generally needs to be handled using a different query mechanism - using `.A_libs` and`.A_all_libs` instead of `.libs` and `.all_libs`.

For me, the compelling case is illustrated in PR #7257 (which I'm working on breaking into smaller, per-package PRs). Namely, some of the changes in the `mfem` package (I'm an mfem developer), such as the [hypre](https://github.com/v-dobrev/spack/blob/9ae4ad573253e54d71105559bca2648236b41dcb/var/spack/repos/builtin/packages/mfem/package.py#L229-L230), [superlu-dist](https://github.com/v-dobrev/spack/blob/9ae4ad573253e54d71105559bca2648236b41dcb/var/spack/repos/builtin/packages/mfem/package.py#L245-L246), [sundials](https://github.com/v-dobrev/spack/blob/9ae4ad573253e54d71105559bca2648236b41dcb/var/spack/repos/builtin/packages/mfem/package.py#L254-L258), etc. dependencies. I'd like to be able to express the mfem dependecies only through its direct dependencies, specifying the component libraries that I need from them (like in the case of [suite-sparse](https://github.com/v-dobrev/spack/blob/9ae4ad573253e54d71105559bca2648236b41dcb/var/spack/repos/builtin/packages/mfem/package.py#L248-L252)). I do not want to go through the dependencies of my direct dependencies to figure out what they need - they should provide that information to me through their `.all_libs` property.

The function `all_libs` you propose solves some of these issues. The problem is that it does not allow packages to change that behavior. I think using a recursive definition of `.all_libs` as in the [mfem propsed modifications](https://github.com/v-dobrev/spack/blob/9ae4ad573253e54d71105559bca2648236b41dcb/var/spack/repos/builtin/packages/mfem/package.py#L384-L411) (which is optional) is not that hard to write if a package-writer wants to be more specific about their library dependencies.

Reply:
I just realized that (in the above discussion) I've been using the wrong signature for spec queries: `spec['x+a+b']` instead of `spec['x:a,b']` - here `a,b` represent query parameters for the libraries `a` and `b` which are part of the package `x`. Sorry about the confusion.

Reply:
@v-dobrev Regarding use cases: the construction `spec['x'].all_libs` isn't necessary in some packages because the build system (or `pkg-config`) resolves dependencies, or `x` is being built as a shared library (so it only needs direct dependency information, which can be provided by the `libs` method).

To speak to your point 1 above: If package `x` is built as a static library, then an operation like `all_libs` is helpful, although it still needs to handle correctly pathologies like circular dependencies, so a dependency graph traversal makes more sense than a deduplicated list. Spack's architecture is geared towards shared libraries (the docs make explicit mention of `RPATH` multiple times -- `RPATH` isn't used by static libraries), which is probably one reason `all_libs`-type operations don't come up much in linking commands. If there's strong desire to enable building packages as static libraries only (no shared libraries), then I could see an argument for more widespread use of `all_libs`. Even then, the pitfalls of using `all_libs` with shared libraries should be documented.

The one case I can think of so far where I've encountered a requirement for secondary dependency information is with `hypre` and `mfem`, and that's only because `hypre` can't be built as a shared library on Darwin/OS X, so all of its dependencies must also be linked explicitly. Looking into this has made me realize that I didn't fix the `mfem` package correctly in #7165 (my fix overlinks `hypre` on any non-Darwin/OS X). Even then, linking `hypre` only requires `lapack` & `blas`, and I don't think it's worth creating an `all_libs` method for a case where the dependency graph traversal is that easy.

Reply:
@goxberry, you keep mentioning circular dependencies - I think spack does not allow such dependencies. That is why specs are represented as DAGs. Am I missing something?

You say only the `hypre` dependency can benefit the mfem spack package - that is not how I see things - see my [proposed updated mfem package](https://github.com/v-dobrev/spack/blob/9ae4ad573253e54d71105559bca2648236b41dcb/var/spack/repos/builtin/packages/mfem/package.py).

Another point: you seem to imply that if I'm linking with a correctly linked shared library, I do not need to link with any of its dependencies. There are certainly cases where that is not sufficient.

Reply:
@v-dobrev I'm mainly concerned about guarding against undesirable consequences (like overlinking, forcing unnecessary recompilation, larger binaries, making shared libraries less swappable). Forcing unnecessary recompilation is particularly relevant if Spack moves forward with plans to store pre-built binaries in the same fashion as [Homebrew](https://brew.sh/); overlinking could force rebuilding many more of these binaries than necessary.

My suggested changes are:

- Traverse the Spack DAG as suggested by @alalazo; conversion to `LibraryList` eliminates degree information contained in the DAG that could be used to compute secondary dependencies of shared libraries.
- Secondary dependencies of shared libraries should be treated differently: for these dependencies, replace `-L` with `-Wl,-rpath-link` or `-Wl,-rpath`, depending on whether binaries should be relocatable, and the `-l` flags of these dependencies should be omitted.

Your comment above that

> The proposed default implementation always returns an overestimate, i.e. a set containing all necessary dependencies plus potentially some redundant dependencies. This may not be optimal but should always work.

is exactly the case I am concerned about, because "may not be optimal" can actually have significant consequences (for instance, [overlinking requiring Debian to rebuild three times as many packages as needed when a new version of freetype was added](https://lists.debian.org/debian-devel-announce/2005/11/msg00016.html)).

> @goxberry, you keep mentioning circular dependencies - I think spack does not allow such dependencies. That is why specs are represented as DAGs. Am I missing something?

Re: circular dependencies, I'm referring to symbol resolution done by a linker (e.g., the case discussed [here](https://eli.thegreenplace.net/2013/07/09/library-order-in-static-linking), which I also linked to above), when `libX` has a function call from `libY`, and `libY` has a function call from `libX`. This concept is separate from the Spack DAG.

> You say only the `hypre` dependency can benefit the mfem spack package - that is not how I see things - see my proposed updated mfem package.

No, what I said was:

>  If package `x` is built as a static library, then an operation like `all_libs` is helpful, although it still needs to handle correctly pathologies like circular dependencies, so a dependency graph traversal makes more sense than a deduplicated list.

and

> The one case I can think of so far where I've encountered a requirement for secondary dependency information is with `hypre` and `mfem`

The `mfem` spack package is built as a static library, and as I mentioned above, this use case is sensible for `all_libs`. 

However, if it were ever built as a shared library in the future, then `mfem~lapack^hypre+shared` would be overlinked because it would directly link to BLAS and LAPACK, even though neither BLAS nor LAPACK is a direct dependency of MFEM. I don't think this issue would be obvious to many users.

> Another point: you seem to imply that if I'm linking with a correctly linked shared library, I do not need to link with any of its dependencies. There are certainly cases where that is not sufficient.

No, what I am saying is that using `-L`/`-l` for secondary dependencies is bad practice when the secondary dependency is a direct dependent of a shared library. In these cases, the `-l` flag should be omitted and the `-L` flag should be replaced with `-Wl,-rpath-link` or `-Wl,-rpath`. [Right now, the `ld_flags` method adds `-L`/`-l` for all libraries in a `LibraryList`, which I am arguing is undesirable for secondary dependencies.](https://github.com/spack/spack/blob/develop/lib/spack/llnl/util/filesystem.py)

Let's say we have executable `A` that depends on `libX`, and `libX` depends on `libY`, and both `libX` and `libY` are shared libraries. Let's also assume `A` does _not_ directly depend on `libY` -- it does not use any symbols, like function calls, from `libY` -- so `libY` is a secondary dependency of `A`. If `libX` wasn't linked to `libY` when `libX` was built, then when `A` is built, the linker must figure out how to find the symbols defined in `libY`. There are a few ways to do so (and this list is not meant to be exhaustive):

- `gcc -o A A.c -L${PATH_TO_X} -lX -L${PATH_TO_Y} -lY` overlinks the library because now `A` contains symbol definitions from `libY`

- `gcc -o A A.c -L${PATH_TO_X} -lX -Wl,-rpath-link=${PATH_TO_Y}` gives the linker the path to `libY` for symbol resolution

- `gcc -o A A.c -L${PATH_TO_X} -lX -Wl,-rpath=${PATH_TO_Y}` linker the path to `libY` for symbol resolution and stores it (hardcoded) in the executable itself, so `libY` can't be relocated, but the `rpath` can be changed after building `A` with a tool like `chrpath`

All of these will yield a functional executable `A`, but if one wants to replace `libY` with a newer version of `libY` (subject to things like ABI compatibility), the first case requires recompiling `A`, the second doesn't, and if the newer `libY` is installed in the same place as the older `libY`, then the third case also doesn't require recompiling `A`. Again, this is all covered in the sources I linked to above.

As you note, secondary dependency information is needed, but there are different ways to pass that information to the linker, and it is possible to do so without including unnecessary symbols into binaries that can force unnecessary recompilation.

Reply:
@goxberry, let me try to express my purpose for this PR:
1. Propose the addition of two new properties for packages: `all_libs` and `all_headers` which allow package-writers to extract link and compile options (respectively) for using a given package including necessary dependencies.
2. Propose a simple implementation of 1.

If you are against 1, there is no need for further discussion.

If you are against 2, i.e. the proposed implementation, then please explain what you are proposing. I really do not care much about 2 but I really would like to have 1. Of course, the names - `all_libs` and `all_headers` are just suggestions - I'm only interested in the functionality.

Reply:
@v-dobrev I support the addition of the two new properties. I support a simple implementation of `all_headers`. I also support a simple implementation of `all_libs` on the condition that it will soon be made more robust to avoid the overlinking issue. The one change I'd suggest now would be to use @alalazo's DAG traversal pattern to implement `all_libs` in place of your proposed `_all_libs_default_handler`.

The more I've thought about `all_libs`, the more I see its utility, and I think it could simplify dramatically Spack packaging, and it could also potentially prevent things like overlinking in the future, because we currently have to guard against that manually.

The main reason I've made so much of a fuss is that this PR changes the Spack core, which affects all packages. As someone who has contributed to package managers for the past 5 years, I've watched users write packages by looking at examples of already written packages. My concern is that users will look at the `all_libs` pattern -- which is very compelling -- and use it widely, with well-meaning, but unintended and hard-to-debug consequences.

Reply:
@goxberry, I'm glad we agree on 1.

I can remove the `all_libs` attribute/property and replace it with a function. However, that way package writers lose the ability to control the behavior of `all_libs` for their package. Isn't that more restrictive? On the other hand, what is the advantage of implementing `all_libs` as a function?

My take on the dilemma: "Do I add a useful feature that can be misused badly?" is that I will usually choose "Yes" over "No". Of course, it is always good to consider the consequences.

Reply:
I guess one reason to prefer @alalazo's proposal is the `specifiers` dictionary.

How about we make `all_libs` a virtual method of the package class (so that it can be overridden) with one optional parameter: `specifiers=None`?

Or, do you want to explicitly prevent packages from overriding the `all_libs` implementation?

Reply:
I think from design perspective ‘all_libs()’ shall not be a virtual method as it postprocesses a DAG of packages and uses ‘libs()’ or alike from each separate package. 

I am not in favor allowing packages to override it anyhow.

Reply:
@davydden, except for "design perspective" and "not in favor", is there any reason to prefer that solution?

Because, on the other side, I have clear use cases that will simplify usage for people writing packages. I can elaborate more if it is not clear what the use cases are.

Reply:
@v-dobrev I think you can achieve everything with what @alalazo suggests.

Reply:
That is correct. All my proposal does is to make it simpler for package developers.

With @alalazo's solution a user of mfem will need to know all dependencies of mfem, including dependencies based on variants. With a library like mfem (and many other libraries) this process is not that simple.

Let me give you an example: let's assume that a user specifies that his package, A, depends on mfem (no special variant requirements). Also, let's assume that the concretization process for A resolves the mfem spec has to be `mfem+netcdf` variant. Now, with @alalazo's proposal, this means that package A has to call `all_libs` and explicitly specify the requirement to link with hdf5, including the `hdf5_hl` library (which is not returned by the `libs` query to `hdf5` unless you ask for it specifically). In what I propose, the person writing package A does not need to worry about details like these. Instead they just query `spec['mfem'].all_libs` (which has to be implemented properly).

Is there a reason to deny this convenience to users?

Reply:
> the person writing package A does not need to worry about details like these. 

it would be exactly like this with @alalazo suggestion, if you have a build system which **need** to know all dependencies, then package developers simply should not use `libs()` but instead that new function to post-process the DAG. In other words, they simply should not assume anything about their dependencies like `mfem`.

Sorry, I don't see what is the advantage.

Reply:
So **convenience for users** is not an advantage?

Reply:
> So convenience for users is not an advantage?

as I said, I am sorry but I don't see how it is more convenient.

Reply:
Maybe I'm not explaining this well.

Compare `spec['mfem'].all_libs` and `all_libs(spec['mfem'], {'hdf5': 'hl'})` - the fact that in the first case the user does not need to know the specific requirements of mfem, compared to the second case where they need to know or figure out by trial and error that (in some cases) they need to ask for `{'hdf5': 'hl'}`. Isn't the first option more convenient?

Reply:
I had to catch up with the conversation here, and I think there are a lot of interesting points on the table. 

My understanding so far is that the use case for `all_libs` is (obviously) static linking. In that sense I think it would be useful to have the possibility to point to a node in the DAG and ask for the list of libraries that are needed to link that node statically.

I still have concerns though on the implementation as a query. With the current implementation:
```python
libs = spec.libs
for dep in spec.dependencies(deptype='link'):
    libs += spec[dep.name].libs # <- here we stop at direct dependencies
￼return libs
```
we won't obtain the full linking line. If we recurse instead:
```python
libs = spec.libs
for dep in spec.dependencies(deptype='link'):
    libs += spec[dep.name].all_libs
￼return libs
```
we'll risk of touching the same node multiple times from different edges. This might be a mess if we add query parameters to the picture (they might be redundant or conflicting).

@v-dobrev objection to what I propose:

> he fact that in the first case the user does not need to know the specific requirements of mfem, compared to the second case where they need to know or figure out by trial and error that (in some cases) they need to ask for {'hdf5': 'hl'}

is a fair point for large DAGs. If we think of something as complex as `boost` or `petsc`, we can't expect a user to maintain an exotic query on all their dependencies at once. 

That said, I still need to figure out what could be the best / neatest implementation, but my take so far is this:

1. We need to traverse the entire DAG, in topological order, at once. In that sense I think the approach I propose is better, including the possibility to have a dictionary of `specifiers`.
2. Probably it's the dictionary of `specifiers` that needs to be **computed locally** based on direct dependencies, and **reduced** among all nodes in the DAG before being passed to something like the `all_libs` function
3. It might be a good idea, if we implement something like this, to be defensive and let it work only if we are building a static configuration 

Thoughts?

Reply:
@alalazo,

I think we can do the construction of the `specifiers` dictionary and compose the result of `all_libs` in one DAG traversal.

I agree that the dictionary `specifiers` may be useful, that is why I proposed to implement `all_libs` as a virtual method with one argument, `specifiers=None`. (I'm not a python expert, but I assume implementing virtual methods is not a problem in python.) I imagine that the default implementation in the base class (`Package`) will look something like this:
```python
def all_libs(self, specifiers=None):
    pkg = self.name
    if pkg in specifiers and specifiers[pkg] is not None:
        pkg += ':' + ','.join(specifiers[pkg])
    libs = self.spec[pkg].libs

    # Comment 1:
    # When packages override `all_libs` they can insert here any additional query
    # parameters in the `specifiers` dictionary.

    for dep in self.spec.dependencies(deptype='link', depth=1):
        # let's assume that the above is how one loops only over the direct,
        # (depth = 1), dependencies of this package.

        # Comment 2:
        # This is where a package can decide if it really needs to link with `dep`.
        # For example, if this package is build as shared and the library is already
        # linked with `dep` - then the line below can be skipped, potentially.
        libs += self.spec[dep].all_libs(specifiers)

    return libs
```
I think this implementation addresses all points 1-3 that you raised, and also allows packages to modify the behavior by inserting code where comments 1 and 2 are (in their implementation of `all_libs`).

What do you think of this approach?

Reply:
In the code above, I did not handle properly virtual packages but, hopefully, the idea is clear: in addition to checking `specifiers[pkg]`, one will need to add any `specifiers[vpkg]` for the virtual packages, `vpkg`, provided by this package.

Reply:
I'll try to find some time today to have a good look at the implementation above.

p.s.  My C++ sense is tingling :wink:
> (I'm not a python expert, but I assume implementing **virtual methods** is not a problem in python.)

 Jokes apart, in python every method is "virtual". What happens when you have an instance of a class in a hierarchy and search for a method or attribute, is that you iterate forward starting from the instance and proceeding upwards to its class and parents of its class. The first entity that can serve the method or attribute will be the one "winning". If you are interested in the details, the whole procedure is called "Method Resolution Order" and usually referred to as MRO.

Reply:
@v-dobrev 

> Compare spec['mfem'].all_libs and all_libs(spec['mfem'], {'hdf5': 'hl'})

Once DAG is conretized, then each node is specified completely and essentially it should be something like `all_libs(self.spec['mfem'])` and that's it. So it's not up to a package which use `mfem` to try to figure out what's the final configuration of `mfem` and whether or not it contains all dependencies or only some dependencies. That information should already be available in DAG.

In other words, I see it as: we have a DAG which is concretized, say packageA==nodeA uses `mfem`, from the DAG any node should be able to query: "give me all the libs I need to statically link with traversed from myself".

But let's wait for @alalazo to comment, he knows internals very well. Perhaps I am missing something in my understanding expressed above.

Reply:
@alalazo: 

> My understanding so far is that the use case for `all_libs` is (obviously) static linking.

I think the use case scope is limited to cases where one wants to generate link lines for all dependencies of a given package within a `MakefilePackage` or `Package`. This information usually isn't necessary for other types of build systems.

Static linking is (almost) the easy case -- there is the minor hiccup of circular dependencies (from the perspective of the linker, not Spack) that require constructions like `-lfoo -lbar -lfoo`. This minor issue isn't a problem with shared libraries on OS X or Linux.

If Spack already takes care of `rpath` resolution for secondary dependencies via environment variables, then there's no need to use `all_libs` for secondary dependencies. Otherwise, it could be useful in cases with shared libraries also, and the degree of DAG nodes with respect to the package being built would help determine which flags should be used. The default `ld_flags` behavior doesn't seem to do the right thing for some secondary dependencies (specifically, in the case where a secondary dependent is a dependent of a direct dependent that is a shared library -- I realize the wording here isn't great).

> 1. We need to traverse the entire DAG, in topological order, at once. In that sense I think the approach I propose is better, including the possibility to have a dictionary of `specifiers`.
> 2. Probably it's the dictionary of `specifiers` that needs to be *computed locally* based on direct dependencies, and *reduced* among all nodes in the DAG before being passed to something like the `all_libs` function
> 3. It might be a good idea, if we implement something like this, to be defensive and let it work only if we are building a static configuration

I agree that traversing the DAG is better -- depth information can be used to compute flags in cases with shared libraries. I also think that being defensive is a good idea until there are better tests that check for over/underlinking in package builds. (On Linux, such tests could look at the output from `ldd`; I haven't seen an OS X equivalent with something like `otool`.)

Reply:
@alalazo, thanks for taking the time to consider this. And thanks for the MRO explanation.

@davydden, I think you are dismissing a common situation where a package builds multiple libraries but there are no separate variants for these libraries. An example is SuiteSparse - it builds `libumf.so`, `libklu.so`, etc. When mfem links with SuiteSparse, it only needs `umf`, `klu` and their dependencies (which is not all the libraries build by SuiteSparse). If you only rely on the DAG, you cannot describe these finer grain dependencies.

@goxberry, I think CMake can use this too - not all CMake packages export their dependencies as CMake package files.

Reply:
@v-dobrev Agreed, CMake and even Autotools sometimes have options for link flags instead of prefixes.

Reply:
@v-dobrev Just wanted to drop a line, saying I am currently swamped with other things - but didn't forget about this. Will probably come back in the week-end. 

OT: are you registered on Spack's slack?

Reply:
I just registered there (Veselin Dobrev).